wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/KubernetesToolToWorkspaceApplier.java

```diff
@@ -88,22 +94,61 @@ public void apply(
             || OPENSHIFT_TOOL_TYPE.equals(k8sTool.getType()),
         format("Plugin must have `%s` or `%s` type", KUBERNETES_TOOL_TYPE, OPENSHIFT_TOOL_TYPE));
 
-    String recipeFileContent = retrieveContent(k8sTool, contentProvider);
-
-    List<HasMetadata> list = unmarshal(k8sTool, recipeFileContent);
+    String toolContent = retrieveContent(k8sTool, contentProvider);
 
+    List<HasMetadata> toolObjects = unmarshalToolObjects(k8sTool, toolContent);
     if (!k8sTool.getSelector().isEmpty()) {
-      list = filter(list, k8sTool.getSelector());
+      toolObjects = filter(toolObjects, k8sTool.getSelector());
     }
 
-    estimateCommandsMachineName(workspaceConfig, k8sTool, list);
+    K8sToolEnvironmentBuilder environmentBuilder =
+        envBuilderFactory.create(k8sTool, workspaceConfig);
+
+    List<PodData> toolPods = removePods(toolObjects);
 
-    RecipeImpl recipe =
-        new RecipeImpl(k8sTool.getType(), YAML_CONTENT_TYPE, asYaml(k8sTool, list), null);
+    environmentBuilder.addRecipeObjects(toolObjects);
+    if (!toolPods.isEmpty()) {
+      environmentBuilder.addPods(toolPods);
+      estimateCommandsMachineName(
+          workspaceConfig, k8sTool, environmentBuilder.getDeployment(), toolPods);
+    }
 
-    String envName = k8sTool.getName();
-    workspaceConfig.getEnvironments().put(envName, new EnvironmentImpl(recipe, emptyMap()));
-    workspaceConfig.setDefaultEnv(envName);
+    workspaceConfig.setDefaultEnv(DEFAULT_ENV_NAME);
+    workspaceConfig.getEnvironments().put(DEFAULT_ENV_NAME, environmentBuilder.buildEnvironment());
+  }
+
+  /**
+   * Removes Deployments/Pods objects from the specified list and return removed items.
+   *
+   * @param objects list of objects where pods should be removed
+   * @return list which contains removed deployments/pods
+   */
+  private List<PodData> removePods(List<HasMetadata> objects) {
+    Iterator<HasMetadata> objectIt = objects.iterator();
+    List<PodData> pods = new ArrayList<>();
+    while (objectIt.hasNext()) {
+      HasMetadata object = objectIt.next();
+      if (object instanceof Deployment) {
+        objectIt.remove();
+
+        Deployment deployment = (Deployment) object;
+        PodTemplateSpec podTemplate = deployment.getSpec().getTemplate();
+        pods.add(
+            new PodData(
+                new PodSpecBuilder(podTemplate.getSpec()).build(),
+                new ObjectMetaBuilder(podTemplate.getMetadata()).build()));
+      } else if (object instanceof Pod) {
+        objectIt.remove();
+
+        Pod pod = (Pod) object;
+        pods.add(
+            new PodData(
+                new PodSpecBuilder(pod.getSpec()).build(),
+                new ObjectMetaBuilder(pod.getMetadata()).build()));
+      }
```

Comment:
Shouldn't we handle also replica sets and/or nested kubernetes lists?

Reply:
Kubernetes list can not contains another lists as items.

Reply:
And about replica sets - K8s/OS infra does not support `replica set` yet